src/nemos/basis/_basis_mixin.py

```diff
@@ -255,4 +489,182 @@ def to_transformer(self) -> TransformerBasis:
         ... )
         >>> gridsearch = gridsearch.fit(X, y)
         """
-        return TransformerBasis(copy.deepcopy(self))
+        return TransformerBasis(self)
+
+
+class CompositeBasisMixin:
+    """Mixin class for composite basis.
+
+    Add overwrites concrete methods or defines abstract methods for composite basis
+    (AdditiveBasis and MultiplicativeBasis).
+    """
+
+    def __init__(self, basis1: Basis, basis2: Basis):
+        # deep copy to avoid changes directly to the 1d basis to be reflected
+        # in the composite basis.
+        self.basis1 = copy.deepcopy(basis1)
+        self.basis2 = copy.deepcopy(basis2)
+
+        # set parents
+        self.basis1._parent = self
+        self.basis2._parent = self
+
+        shapes = (
+            *(bas1._input_shape_ for bas1 in basis1._iterate_over_components()),
+            *(bas2._input_shape_ for bas2 in basis2._iterate_over_components()),
+        )
+        # if all bases where set, then set input for composition.
+        set_bases = [s is not None for s in shapes]
+
+        if all(set_bases):
+            # pass down the input shapes
+            self.set_input_shape(*shapes)
+        elif any(set_bases):
+            warnings.warn(
```

Comment:
this isn't actually necessary, the following runs fine. should it?
```python
b = nmo.basis.BSplineEval(5)
b.setup_basis(np.ones((100, 5)))
b2 = nmo.basis.BSplineEval(5)
c = b+b2
c.compute_features(np.ones((100, 5)), np.ones((100,2)))

```

Reply:
this always works, but the transformer version no, I can move the warning to where it is more appropriate

Reply:
actually, whatever, if one tries to use the transformer it will have an error and then figure out that the shape was not initialized